### PR TITLE
Backend models update

### PR DESCRIPTION
**Pull request by dogpineapple**
The current models are outdated and should be similar to our new database schema. While making changes to these models, the `routes` and `tests` that rely on these models had to be updated as well. (Without changes, they were broken and failed tests.)

~~The `Comment` model and `/routes/comments` route has not been updated yet since we won't be focusing on these functionalities. But these can be easily updated later. (Expect `comment.test.ts` to fail)~~ **nvm, it's updated now 8)**

### Backend changes
- Added `/models/user_auth.ts` file for the `UserAuth` class. The `authentication` and `register` methods have been migrated from the `User` model to this new `UserAuth` class. 
- Renamed `/models/favorite.ts` to `/models/bookmark.ts` and the `Favorite` class to `Bookmark`. The rename was necessary since we are renaming the `favorites` table to `bookmarks`. 
- Updated SQL queries in a number of class's methods within the `models/` directory. This was due to the new columns we introduced in the new database schema.
- Updated `data.sql` to fit new database schema
- Updated the `user`, `userAuth`, `usersRoutes`, `postsRoutes`, `post`, `bookmark`, `bookmarksRoutes` tests due to the changes in the models and routes. 
- Added a `membershipEligibility.ts` file to store constants for the 4 different states of a user's `membership_eligibility`. 
- Converted `BCRYPT_WORK_FACTOR` constant in the `config.ts` file to a number. 

### Tests ran
- All BE tests (EXCEPT FOR THE COMMENTS) passed.

### CODE REVIEW

File: bloggies_backend/src/routes/bookmarks.ts [hunk 1]
```diff
@@ -1,42 +1,42 @@
 import { NextFunction, Request, Response } from "express";
 import { ensureLoggedIn } from "../middleware/auth";
 import express from "express";
-import Favorite from "../models/favorite";
+import Favorite from "../models/bookmarks";
```
Comment (megmcdevitt) [inline bloggies_backend/src/routes/bookmarks.ts]: Tiny thing here, but the file is "bookmark.ts" (singular). Got a TS error! (Or change file name to plural)

Also on this line, change Favorite to Bookmark

File: bloggies_backend/src/data.sql [hunk 2]
```diff
@@ -11,25 +11,40 @@ CREATE DATABASE "learning_circle";
```
Comment (megmcdevitt) [inline bloggies_backend/src/data.sql:11]: Check out Line 7 - change to "bookmarks" :)

File: bloggies_backend/src/__tests__/bookmark.test.ts [hunk 3]
```diff
@@ -0,0 +1,122 @@
+import db from "../db";
+import Bookmark from "../models/bookmarks";
```
Comment (megmcdevitt) [inline bloggies_backend/src/__tests__/bookmark.test.ts]: This also should be "bookmark" (singular)

File: bloggies_backend/src/__tests__/bookmark.test.ts [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+let validUserId: number;
+
+const validUserEmail = 'bookmarks@test.com';
+const validUserDisplayName = 'bookmarksuser';
+const validPostDescription = 'Made with Strawberry, Basil, Sparkling Water';
+const validPostTitle = 'Strawberry Basil Soda';
+
+/** Tests for Favorite model methods*/
+describe("Test favorite class model", function () {
+
```
Comment (megmcdevitt) [inline bloggies_backend/src/__tests__/bookmark.test.ts:14]: Line 12 and 13, change to "bookmark" from "favorite"

File: bloggies_backend/src/__tests__/bookmark.test.ts [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+        FROM bookmarks AS b 
+        JOIN posts AS p ON p.id = b.post_id 
+        JOIN users AS u ON u.user_id = b.user_id
+        WHERE p.id = $1 AND u.user_id = $2`,
+      [validPostId, validUserId]);
+
+    const favorite = expectedRes.rows[0];
+    expect(favorite.title).toBe(validPostTitle);
+    expect(favorite.display_name).toBe(validUserDisplayName);
+  });
```
Comment (megmcdevitt) [inline bloggies_backend/src/__tests__/bookmark.test.ts:60]: Lines 57-59 change to "bookmark"

File: bloggies_backend/src/__tests__/bookmarksRoutes.test.ts [hunk 6]
```diff
@@ -37,44 +39,44 @@ describe("Test Post routes", function () {
...
-    await db.query(`INSERT INTO favorites(post_id, user_id) VALUES (${validPostId}, ${validUserId})`);
+    await db.query(`INSERT INTO bookmarks(post_id, user_id) VALUES (${validPostId}, ${validUserId})`);
   });
 
-  /** GET /favorites/:uid => status 200, { posts }.**/
-  test("GET /favorites/:uid - retrieve favorited posts for a user by user id", async function () {
-    const resp = await request(app).get(`/favorites/${validUserId}`);
+  /** GET /bookmarks/:uid => status 200, { posts }.**/
+  test("GET /bookmarks/:uid - retrieve favorited posts for a user by user id", async function () {
+    const resp = await request(app).get(`/bookmarks/${validUserId}`);
```
Comment (megmcdevitt) [inline bloggies_backend/src/__tests__/bookmarksRoutes.test.ts:47]: Not super critical, but the descriptions in this file can also be changed from "favorites" to "bookmarks"

File: bloggies_backend/src/routes/bookmarks.ts [hunk 7]
```diff
@@ -1,42 +1,42 @@
...
 
-/** GET /favorites/:uid - retrieve favorited posts for a user.
+/** GET /bookmarks/:uid - retrieve favorited posts for a user.
  * Returns a list of posts */
-favoritesRouter.get("/:uid", async function (req: Request, res: Response, next: NextFunction) {
+bookmarksRouter.get("/:uid", async function (req: Request, res: Response, next: NextFunction) {
   const userId = parseInt(req.params.uid);
   try {
-    const posts = await Favorite.getAllFavorites(userId);
+    const posts = await Favorite.getAllBookmarks(userId);
```
Comment (megmcdevitt) [inline bloggies_backend/src/routes/bookmarks.ts]: Favorite.getAllBookmarks => Bookmark.getAllBookmarks

(same for more on this page)

**Review by megmcdevitt — CHANGES_REQUESTED** [review]
This looks mostly good! There are some "favorite" remnants left that I pointed out. Searching for "favorite" helped identify them! Some are not critical, but it would be good to have a clean sweep.

Also, the route is called "bookmarks.ts" and the model is "bookmark.ts" - would be good to pick one for cohesion (some of my comments allude to this).

**Review by abquansah — CHANGES_REQUESTED** [review]
It looks like in the file `src/models/comments`, on line 15, you are referencing the column of `id` in the `users` table, when it's been changed to `user_id`, which threw an error for me when I was trying to query the end point earlier. I would do a double check through all the files to ensure all references are updated to the correct columns since we changed the schemas. It looks like on line 36 of the same file, it would be the same issue.

**Comment by dogpineapple** [comment]
> It looks like in the file `src/models/comments`, on line 15, you are referencing the column of `id` in the `users` table, when it's been changed to `user_id`, which threw an error for me when I was trying to query the end point earlier. I would do a double check through all the files to ensure all references are updated to the correct columns since we changed the schemas. It looks like on line 36 of the same file, it would be the same issue.

Got it! It's updated now :D

**Review by megmcdevitt — APPROVED** [review]
Looks good to me!

**Review by abquansah — APPROVED** [review]
Same for me, I'm assuming for the other routes that might need to be added we will add them soon, so this is fine!

**Pull request merged**